src/java/edu/ucsd/library/dams/api/DAMSAPIServlet.java

```diff
@@ -3733,28 +3723,55 @@ private String indexQueue( String objid, String type, int priority )
 		String error = null;
 		if ( queueEnabled && queueSession != null )
 		{
-			try
-			{
-				TextMessage msg = queueSession.createTextMessage(
-					"DAMS Queue Message: " + objid + " (" + type + ")"
-				);
-				msg.setStringProperty("pid",objid);
-				msg.setStringProperty("methodName",type);
+			int numOfTry = 0;
 
-				if ( priority < 1 || priority > 9 )
+			do
+			{
+				try
 				{
-					priority = DEFAULT_PRIORITY;
+					TextMessage msg = queueSession.createTextMessage(
+						"DAMS Queue Message: " + objid + " (" + type + ")"
+					);
+					msg.setStringProperty("pid",objid);
+					msg.setStringProperty("methodName",type);
+	
+					if ( priority < 1 || priority > 9 )
+					{
+						priority = DEFAULT_PRIORITY;
+					}
+					queueProducer.send(
+						msg, queueProducer.getDeliveryMode(),
+						priority, queueProducer.getTimeToLive()
+					);
+	
+					error = "";
 				}
-				queueProducer.send(
-					msg, queueProducer.getDeliveryMode(),
-					priority, queueProducer.getTimeToLive()
-				);
-			}
-			catch ( Exception ex )
-			{
-				log.error("Error sending event to queue", ex );
-				error = "Error sending object to queue: " + ex.toString();
-			}
+				catch ( Exception ex )
+				{
+					log.error( "Error sending event to queue", ex );
+					error = "Error sending object to queue: " + ex.toString();
+	
+					if (queueSession != null)
+					{
+					  try
+					  {
+						  queueSession.close();
+						  queueConnection.close();
+					  } catch (JMSException e) {}
+					}
+	
+					try
+					{
+						// re-initialize the JMS session.
+						initJmsSession(queueUrl, queueName);
+					}
+					catch (JMSException e)
+					{
+						log.error( "Error sending event to queue", e );
+						error = "Error sending object to queue: " + e.toString();
+					}
+				}
+			} while (++numOfTry < MAX_INDEX_TRY);
```

Comment:
This code seems to run the `do/while` loop twice regardless of success with the first run through the `try` block. Is that correct? And if so, is that what we want? I suppose I would have imagined that upon a successful message send where no error is caught, the do/while loop would exit. But I may be reading this wrong.

Reply:
@mcritchlow Good catch!